pubsub/awssnssqs/awssnssqs.go

```diff
@@ -589,38 +622,38 @@ func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*dr
 	}
 	var ms []*driver.Message
 	for _, m := range output.Messages {
-		// By default, messages from SNS come with everything in a JSON-encoded Body.
-		// For example, message attributes are not in m.MessageAttributes, but
-		// are rather in a MessageAttributes field in the JSON-encoded body.
-		// If RawMessageDelivery is enabled as part of the SQS queue subscription
-		// to the SNS topic, then body is in Body and attributes are in
-		// MessageAttributes. Same deal if you send directly to the SQS queue.
-		//
-		// If it looks like the body might be JSON, we try decoding it. If it
-		// doesn't look like JSON, or if the JSON decode fails, we use the raw
-		// data from m.
 		bodyStr := aws.StringValue(m.Body)
 		rawAttrs := map[string]string{}
-		useJSON := len(m.MessageAttributes) == 0 // if we got attributes, it's raw
-		if useJSON {
+
+		// If the user told us that message bodies are raw, or if there are
+		// top-level MessageAttributes, then it's raw.
+		raw := s.opts.Raw || len(m.MessageAttributes) > 0
```

Comment:
We can't rely on it, because any message can have 0 attributes, but if there **are** top-level attributes, then we know for sure it is raw. SNS JSON messages don't have any top-level attributes; the attributes are encoded in the JSON.

Reply:
(extended the comment to make this more explicit)